mempool/mempool.go

```diff
@@ -0,0 +1,142 @@
+	// maxGas.
+	// If both maxes are negative, there is no cap on the size of all returned
+	// transactions (~ all available transactions).
+	ReapMaxBytesMaxGas(maxBytes, maxGas int64) types.Txs
+
+	// ReapMaxTxs reaps up to max transactions from the mempool.
+	// If max is negative, there is no cap on the size of all returned
+	// transactions (~ all available transactions).
+	ReapMaxTxs(max int) types.Txs
+
+	// Lock locks the mempool. The consensus must be able to hold lock to safely update.
+	Lock()
+
+	// Unlock unlocks the mempool.
+	Unlock()
+
+	// Update informs the mempool that the given txs were committed and can be discarded.
+	// NOTE: this should be called *after* block is committed by consensus.
+	// NOTE: Lock/Unlock must be managed by caller
+	Update(
+		blockHeight int64,
+		blockTxs types.Txs,
+		deliverTxResponses []*abci.ResponseDeliverTx,
+		newPreFn PreCheckFunc,
+		newPostFn PostCheckFunc,
+	) error
+
+	// FlushAppConn flushes the mempool connection to ensure async reqResCb calls are
+	// done. E.g. from CheckTx.
+	// NOTE: Lock/Unlock must be managed by caller
+	FlushAppConn() error
+
+	// Flush removes all transactions from the mempool and cache
+	Flush()
+
+	// TxsAvailable returns a channel which fires once for every height,
+	// and only when transactions are available in the mempool.
+	// NOTE: the returned channel may be nil if EnableTxsAvailable was not called.
+	TxsAvailable() <-chan struct{}
+
+	// EnableTxsAvailable initializes the TxsAvailable channel, ensuring it will
+	// trigger once every height when transactions are available.
+	EnableTxsAvailable()
+
+	// Size returns the number of transactions in the mempool.
+	Size() int
+
+	// TxsBytes returns the total size of all txs in the mempool.
+	TxsBytes() int64
+
+	// InitWAL creates a directory for the WAL file and opens a file itself. If
+	// there is an error, it will be of type *PathError.
+	InitWAL() error
+
+	// CloseWAL closes and discards the underlying WAL file.
+	// Any further writes will not be relayed to disk.
+	CloseWAL()
+}
+
+//--------------------------------------------------------------------------------
+
+// PreCheckFunc is an optional filter executed before CheckTx and rejects
+// transaction if false is returned. An example would be to ensure that a
+// transaction doesn't exceeded the block size.
+type PreCheckFunc func(types.Tx) error
+
+// PostCheckFunc is an optional filter executed after CheckTx and rejects
+// transaction if false is returned. An example would be to ensure a
+// transaction doesn't require more gas than available for the block.
+type PostCheckFunc func(types.Tx, *abci.ResponseCheckTx) error
+
+// TxInfo are parameters that get passed when attempting to add a tx to the
+// mempool.
+type TxInfo struct {
+	// SenderID is the internal peer ID used in the mempool to identify the
+	// sender, storing 2 bytes with each tx instead of 20 bytes for the p2p.ID.
+	SenderID uint16
```

Comment:
Ah OK, this also makes this: https://github.com/lazyledger/optimint/pull/34#discussion_r612681302 obsolete, right?

Reply:
Exactly